engine/src/main/java/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java

```diff
@@ -0,0 +1,691 @@
+          putToReadOnlyCache( db, rows );
+        }
+      }
+    } catch ( Exception e ) {
+      throw new KettleException( e );
+    } finally {
+      if ( db != null ) {
+        db.disconnect();
+      }
+    }
+  }
+
+  private void putToDefaultCache( Database db, List<Object[]> rows ) {
+    final int keysAmount = meta.getStreamKeyField1().length;
+    RowMetaInterface prototype = copyValueMetasFrom( db.getReturnRowMeta(), keysAmount );
+
+    // Copy the data into 2 parts: key and value...
+    //
+    for ( Object[] row : rows ) {
+      int index = 0;
+      // not sure it is efficient to re-create the same on every row,
+      // but this was done earlier, so I'm keeping this behaviour
+      RowMetaInterface keyMeta = prototype.clone();
+      Object[] keyData = new Object[ keysAmount ];
+      for ( int i = 0; i < keysAmount; i++ ) {
+        keyData[ i ] = row[ index++ ];
+      }
+      // RowMeta valueMeta = new RowMeta();
+      Object[] valueData = new Object[ data.returnMeta.size() ];
+      for ( int i = 0; i < data.returnMeta.size(); i++ ) {
+        valueData[ i ] = row[ index++ ];
+        // valueMeta.addValueMeta(returnRowMeta.getValueMeta(index++));
+      }
+      // Store the data...
+      //
+      data.cache.storeRowInCache( meta, keyMeta, keyData, valueData );
+      incrementLinesInput();
+    }
+  }
+
+  private RowMetaInterface copyValueMetasFrom( RowMetaInterface source, int n ) {
+    RowMeta result = new RowMeta();
+    for ( int i = 0; i < n; i++ ) {
+      // don't need cloning here,
+      // because value metas will be cloned during iterating through rows
+      result.addValueMeta( source.getValueMeta( i ) );
+    }
+    return result;
+  }
+
+  private void putToReadOnlyCache( Database db, List<Object[]> rows ) {
+    ReadAllCache.Builder cacheBuilder = new ReadAllCache.Builder( data, rows.size() );
+
+    // all keys have the same row meta,
+    // it is useless to re-create it each time
+    RowMetaInterface returnRowMeta = db.getReturnRowMeta();
+    cacheBuilder.setKeysMeta( returnRowMeta.clone() );
+
+    final int keysAmount = meta.getStreamKeyField1().length;
+    // Copy the data into 2 parts: key and value...
+    //
+    final int valuesAmount = data.returnMeta.size();
+    for ( Object[] row : rows ) {
+      Object[] keyData = new Object[ keysAmount ];
+      System.arraycopy( row, 0, keyData, 0, keysAmount );
+
+      Object[] valueData = new Object[ valuesAmount ];
+      System.arraycopy( row, keysAmount, valueData, 0, valuesAmount );
+
+      cacheBuilder.add( keyData, valueData );
+      incrementLinesInput();
+    }
+    data.cache = cacheBuilder.build();
+  }
+
+  /**
+   * Stop the running query
+   */
+  @Override
+  public void stopRunning( StepMetaInterface smi, StepDataInterface sdi ) throws KettleException {
+    meta = (DatabaseLookupMeta) smi;
+    data = (DatabaseLookupData) sdi;
+
+    if ( data.db != null && !data.isCanceled ) {
+      synchronized ( data.db ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object data.db synchronized on line 599 should not be re-assigned in the same synchronized block on line 600. After this re-assignment, you lose the reference to the synchronized object.